ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,665 @@
+    )
+
+  let mark_finished ~(span : Span.t) =
+    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+        let key = span.context.trace_id in
+        match Hashtbl.find_opt spans key with
+        | None ->
+            debug "Span does not exist or already finished"
+        | Some span_list -> (
+          match
+            List.filter (fun x -> x.Span.context <> span.context) span_list
+          with
+          | [] ->
+              Hashtbl.remove spans key
+          | filtered_list -> (
+              Hashtbl.replace spans key filtered_list ;
+              match Hashtbl.find_opt finished_spans key with
+              | None ->
+                  if Hashtbl.length finished_spans < !max_traces then
+                    Hashtbl.add finished_spans key [span]
+              | Some span_list ->
+                  if List.length span_list < !max_spans then
+                    Hashtbl.replace finished_spans key (span :: span_list)
+            )
+        )
+    )
+
+  let assert_finished x =
+    match x with
+    | None ->
+        false
+    | Some (span : Span.t) -> (
+      match Hashtbl.find_opt finished_spans span.context.trace_id with
+      | None ->
+          false
+      | Some span_list ->
+          List.exists (fun x -> x = span) span_list
+    )
+
+  let since () =
+    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+        let copy = Hashtbl.copy finished_spans in
+        Hashtbl.clear finished_spans ;
+        copy
+    )
+
+  module GC = struct
+    let lock = Mutex.create ()
+
+    let span_timeout = ref 86400.
+
+    let span_timeout_thread = ref None
+
+    let gc_inactive_spans () =
+      Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+          Hashtbl.iter
+            (fun _ (spanlist : Span.t list) ->
+              List.iter
+                (fun (span : Span.t) ->
+                  let elapsed = Unix.gettimeofday () -. span.begin_time in
+                  if elapsed > !span_timeout *. 1000000. then
+                    debug "Tracing: Span %s timed out, forcibly finishing now"
+                      span.context.span_id ;
+                  let span =
+                    Span.finish ~span
+                      ~tags:
+                        [("gc_inactive_span_timeout", string_of_float elapsed)]
+                      ()
+                  in
+                  mark_finished ~span
+                )
+                spanlist
+            )
+            spans
+      )
+
+    let initialise_thread ~timeout =
+      span_timeout := timeout ;
+      span_timeout_thread :=
+        Some
+          (Thread.create
+             (fun () ->
+               while true do
+                 debug "Tracing: Span garbage collector" ;
+                 Thread.delay !span_timeout ;
+                 gc_inactive_spans ()
+               done
+             )
+             ()
+          )
+  end
+end
+
+module Tracer = struct
+  type t = {name: string; provider: provider_config_t ref}
```

Comment:
[ OpenTelemetry specification
](https://opentelemetry.io/docs/reference/specification/trace/api/#tracerprovider-operations)

> Note: This could, for example, be implemented by storing any mutable configuration in the TracerProvider and having Tracer implementation objects have a reference to the TracerProvider from which they were obtained. If configuration must be stored per-tracer (such as disabling a certain tracer), the tracer could, for example, do a look-up with its identity in a map in the TracerProvider, or the TracerProvider could maintain a registry of all returned Tracers and actively update their configuration if it changes.

 If it were about functionality I would refactor Tracer and TracerProvider into one module.

Reply:
Resolved by bbcece5ba2b97f9fcb187f193706b5fb2a364e53, 9e0e83cfa9b9d472a3947e271eb1b961143e4c17

Reply:
I had accidently forgotten some changes I have fixed that now.